go-controller/pkg/ovn/egressfirewall.go

```diff
@@ -0,0 +1,304 @@
+	name        string
+	namespace   string
+	egressRules []*egressFirewallRule
+}
+
+type egressFirewallRule struct {
+	id     int
+	access egressfirewallapi.EgressFirewallRuleType
+	ports  []*port
+	to     destination
+}
+
+type port struct {
+	protocol *string
+	portNum  *int32
+}
+
+type destination struct {
+	cidrSelector string
+}
+
+func newEgressFirewall(egressFirewallPolicy *egressfirewallapi.EgressFirewall) *egressFirewall {
+	ef := &egressFirewall{
+		name:        egressFirewallPolicy.Name,
+		namespace:   egressFirewallPolicy.Namespace,
+		egressRules: make([]*egressFirewallRule, 0),
+	}
+	return ef
+}
+
+func newEgressFirewallRule(rawEgressFirewallRule egressfirewallapi.EgressFirewallRule, id int) (*egressFirewallRule, error) {
+	efr := &egressFirewallRule{
+		id:     id,
+		access: rawEgressFirewallRule.Type,
+		ports:  make([]*port, 0),
+	}
+
+	if rawEgressFirewallRule.To.CIDRSelector == "" {
+		return nil, fmt.Errorf("EgressFirewallRule must have a CIDRselector set")
+	}
+
+	_, _, err := net.ParseCIDR(rawEgressFirewallRule.To.CIDRSelector)
+	if err != nil {
+		return nil, err
+
+	}
+	efr.to.cidrSelector = rawEgressFirewallRule.To.CIDRSelector
+
+	for _, rawPort := range rawEgressFirewallRule.Ports {
+		port := port{}
+		if rawPort.Protocol != nil {
+			port.protocol = rawPort.Protocol
+		}
+		if rawPort.PortNum != nil {
+			port.portNum = rawPort.PortNum
+		}
+		efr.ports = append(efr.ports, &port)
+	}
+
+	return efr, nil
+}
+
+func (oc *Controller) addEgressFirewall(egressFirewall *egressfirewallapi.EgressFirewall) {
+	klog.Infof("Adding egress Firewall %s in namesapce %s", egressFirewall.Name, egressFirewall.Namespace)
+	nsInfo, err := oc.waitForNamespaceLocked(egressFirewall.Namespace)
+	if err != nil {
+		klog.Errorf("Failed to wait for namespace %s event (%v)",
+			egressFirewall.Namespace, err)
+		return
+	}
+	defer nsInfo.Unlock()
+
+	if nsInfo.egressFirewall {
+		klog.Errorf("Attempting to add egressFirewall %s to namespace %s when it already has an egressFirewall",
+			egressFirewall.Name, egressFirewall.Namespace)
+		return
+	}
+
+	ef := newEgressFirewall(egressFirewall)
+	nsInfo.egressFirewallPolicy = ef
+	//lock the newgressFirewall and unlock nsInfo
+	for i, egressFirewallRule := range egressFirewall.Spec.Egress {
+		//process Rules into egressFirewallRules for egressFirewall struct
+		efr, err := newEgressFirewallRule(egressFirewallRule, i)
+		if err != nil {
+			klog.Errorf("Cannot create EgressFirewall Rule for destination %s to namespace %s - %v",
+				egressFirewallRule.To.CIDRSelector, egressFirewall.Namespace, err)
+			continue
+
+		}
+		ef.egressRules = append(ef.egressRules, efr)
+	}
+
+	existingNodes, err := oc.kube.GetNodes()
+	if err != nil {
+		klog.Errorf("Unable to add egressfirewall %s, cannot list nodes: %s", egressFirewall.Name, err)
+	}
+	for _, node := range existingNodes.Items {
+		joinSwitch := joinSwitch(node.Name)
+		err = ef.addACLToJoinSwitch(joinSwitch, nsInfo.addressSet.GetIPv4HashName())
```

Comment:
You need to check IPv4 / IPv6 here, you're again only using `GetIPv4HashName`

Reply:
done